File: java/leetcode_grind/Day343.java
package leetcode_grind;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.PriorityQueue;
import java.util.TreeMap;

public class Day343 {
    // https://leetcode.com/problems/constrained-subsequence-sum/description
    static class Solution {
        public int constrainedSubsetSumHeap(int[] nums, int k) {
            var heap = new PriorityQueue<int[]>((a, b) -> {
                return b[0] - a[0];
            });

            heap.add(new int[] { nums[0], 0 });
            var ans = nums[0];

            for (var i = 1; i < nums.length; i++) {
                while (i - heap.peek()[1] > k) {
                    heap.remove();
                }

                var curr = Math.max(0, heap.peek()[0]) + nums[i];
                ans = Math.max(ans, curr);
                heap.add(new int[] { curr, i });
            }

            return ans;
        }

        public int constrainedSubsetSumTreeMap(int[] nums, int k) {
            TreeMap<Integer, Integer> window = new TreeMap<>();
            window.put(0, 0);

            int[] dp = new int[nums.length];

            for (int i = 0; i < nums.length; i++) {
                dp[i] = nums[i] + window.lastKey();
                window.put(dp[i], window.getOrDefault(dp[i], 0) + 1);

                if (i >= k) {
                    window.put(dp[i - k], window.get(dp[i - k]) - 1);
                    if (window.get(dp[i - k]) == 0) {
                        window.remove(dp[i - k]);
                    }
                }
            }

            int ans = Integer.MIN_VALUE;
            for (int num : dp) {
                ans = Math.max(ans, num);
            }
            return ans;
        }

        public int constrainedSubsetSumDeque(int[] nums, int k) {
            Deque<Integer> deque = new ArrayDeque<>();
            var dp = new int[nums.length];

            for (int i = 0; i < nums.length; i++) {
                if (!deque.isEmpty() && i - deque.peek() > k) {
                    deque.poll();
                }

                dp[i] = (deque.isEmpty() ? 0 : dp[deque.peek()]) + nums[i];

                while (!deque.isEmpty() && dp[deque.peekLast()] < dp[i]) {
                    deque.pollLast();
                }

                if (dp[i] > 0) {
                    deque.offer(i);
                }
            }

            var ans = Integer.MIN_VALUE;
            for (var v : dp) {
                ans = Math.max(ans, v);
            }
            return ans;
        }
    }

    // https://leetcode.com/problems/maximum-subarray/description/
    static class Solution2 {
        public int maxSubArray(int[] nums) {
            var maxSubArray = nums[0];
            var currentSubarray = nums[0];

            for (int i = 1; i < nums.length; i++) {
                int num = nums[i];
                currentSubarray = Math.max(num, currentSubarray + num);
                maxSubArray = Math.max(maxSubArray, currentSubarray);
            }

            return maxSubArray;
        }
    }

    // https://leetcode.com/problems/remove-element/description/
    static class Solution3 {
        public int removeElementForward(int[] nums, int val) {
            var left = 0;
            for (var right = 0; right < nums.length; right++) {
                if (nums[right] == val) {
                    continue;
                }
                nums[left++] = nums[right];
            }
            return left;
        }

        public int removeElementBackwards(int[] nums, int val) {
            var i = 0;
            var n = nums.length;
            while (i < n) {
                if (nums[i] == val) {
                    nums[i] = nums[n - 1];
                    n--;
                } else {
                    i++;
                }
            }
            return i;
        }
    }

    // https://leetcode.com/problems/remove-duplicates-from-sorted-array/description/
    static class Solution4 {
        public int removeDuplicates(int[] nums) {
            var left = 0;
            for (var right = 1; right < nums.length; right++) {
                if (nums[left] != nums[right]) {
                    nums[++left] = nums[right];
                }
            }
            return ++left;
        }

        public int removeDuplicatesLeetcode(int[] nums) {
            var insertIndex = 1;
            for (int i = 1; i < nums.length; i++) {
                if (nums[i - 1] != nums[i]) {
                    nums[insertIndex++] = nums[i];
                }
            }
            return insertIndex;
        }
    }

    // https://leetcode.com/problems/move-zeroes/description/
    static class Solution5 {
        public void moveZeroes(int[] nums) {
            var left = 0;
            for (var right = 0; right < nums.length; right++) {
                if (nums[right] != 0) {
                    var tmp = nums[right];
                    nums[right] = nums[left];
                    nums[left] = tmp;
                    left++;
                }
            }
        }
    }
}

File: src/leetcode_grind/day_343.rs
// https://leetcode.com/problems/constrained-subsequence-sum/description
pub fn constrained_subset_sum_deque(nums: Vec<i32>, k: i32) -> i32 {
    use std::collections::VecDeque;
    let mut queue = VecDeque::new();
    let mut dp = vec![0; nums.len()];

    for i in 0..nums.len() {
        if !queue.is_empty() && i - queue[0] > k as usize {
            queue.pop_front();
        }

        dp[i] = if !queue.is_empty() { dp[queue[0]] } else { 0 } + nums[i];

        while !queue.is_empty() && dp[queue[queue.len() - 1]] < dp[i] {
            queue.pop_back();
        }

        if dp[i] > 0 {
            queue.push_back(i);
        }
    }

    let mut ans = i32::MIN;
    for num in dp {
        ans = ans.max(num);
    }
    ans
}

pub fn constrained_subset_sum_heap(nums: Vec<i32>, k: i32) -> i32 {
    use std::collections::BinaryHeap;
    let mut heap = BinaryHeap::new();
    heap.push((nums[0], 0));

    let mut ans = 0;

    for i in 1..nums.len() {
        while i - heap.peek().unwrap().1 > k as usize {
            heap.pop();
        }

        let curr = 0.max(heap.peek().unwrap().0) + nums[i];
        heap.push((curr, i));
        ans = ans.max(curr);
    }

    ans
}

pub fn constrained_subset_sum_tree_map(nums: Vec<i32>, k: i32) -> i32 {
    use std::collections::BTreeMap;

    let mut window = BTreeMap::new();
    window.insert(0, 0);

    let mut dp = vec![0; nums.len()];
    for i in 0..nums.len() {
        dp[i] = *window.iter().last().unwrap().0 + nums[i];
        *window.entry(dp[i]).or_insert(0) += 1;

        if i >= k as usize {
            let e = window.entry(dp[i - k as usize]).or_insert(0);
            *e -= 1;
            if *e == 0 {
                window.remove(&dp[i - k as usize]);
            }
        }
    }
    dp.into_iter().max().unwrap()
}

// https://leetcode.com/problems/maximum-subarray/description/
pub fn max_sub_array(nums: Vec<i32>) -> i32 {
    let mut max_sub_array = nums[0];
    let mut running_sub_array = nums[0];
    for i in 1..nums.len() {
        running_sub_array = (running_sub_array + nums[i]).max(nums[i]);
        max_sub_array = max_sub_array.max(running_sub_array);
    }
    return max_sub_array.max(running_sub_array);
}
